objectivec/src/ort_session.mm

```diff
@@ -36,81 +39,205 @@ - (nullable instancetype)initWithEnv:(ORTEnv*)env
+  std::vector<const OrtValue*> inputValues;
+  std::vector<OrtValue*> outputValues;
 
-    for (NSString* inputName in inputs) {
-      inputNames.push_back(inputName.UTF8String);
-      inputValues.push_back(static_cast<const OrtValue*>([inputs[inputName] CXXAPIOrtValue]));
-    }
+  for (NSString* inputName in inputs) {
+    inputNames.push_back(inputName.UTF8String);
+    inputValues.push_back(static_cast<const OrtValue*>([inputs[inputName] CXXAPIOrtValue]));
+  }
 
-    for (NSString* outputName in outputNameArray) {
-      outputNames.push_back(outputName.UTF8String);
-      outputValues.push_back(nullptr);
-    }
+  for (NSString* outputName in outputNameArray) {
+    outputNames.push_back(outputName.UTF8String);
+    outputValues.push_back(nullptr);
+  }
 
-    Ort::ThrowOnError(Ort::GetApi().Run(*_session, runOptions,
-                                        inputNames.data(), inputValues.data(), inputNames.size(),
-                                        outputNames.data(), outputNames.size(), outputValues.data()));
-
-    NSMutableDictionary<NSString*, ORTValue*>* outputs = [[NSMutableDictionary alloc] init];
-    for (NSUInteger i = 0; i < outputNameArray.count; ++i) {
-      ORTValue* outputValue = [[ORTValue alloc] initWithCAPIOrtValue:outputValues[i] externalTensorData:nil error:error];
-      if (!outputValue) {
-        // clean up remaining C API OrtValues which haven't been wrapped by an ORTValue yet
-        for (NSUInteger j = i; j < outputNameArray.count; ++j) {
-          Ort::GetApi().ReleaseValue(outputValues[j]);
-        }
-        return nil;
+  Ort::ThrowOnError(Ort::GetApi().Run(*_session, [runOptions CXXAPIOrtRunOptions],
+                                      inputNames.data(), inputValues.data(), inputNames.size(),
+                                      outputNames.data(), outputNames.size(), outputValues.data()));
+
+  NSMutableDictionary<NSString*, ORTValue*>* outputs = [[NSMutableDictionary alloc] init];
+  for (NSUInteger i = 0; i < outputNameArray.count; ++i) {
+    ORTValue* outputValue = [[ORTValue alloc] initWithCAPIOrtValue:outputValues[i] externalTensorData:nil error:error];
+    if (!outputValue) {
+      // clean up remaining C API OrtValues which haven't been wrapped by an ORTValue yet
+      for (NSUInteger j = i; j < outputNameArray.count; ++j) {
+        Ort::GetApi().ReleaseValue(outputValues[j]);
       }
+      return nil;
+    }
+
+    outputs[outputNameArray[i]] = outputValue;
+  }
+
+  return outputs;
+
+  ORT_OBJC_API_IMPL_END_NULLABLE(error)
+}
+
+@end
+
+@implementation ORTSessionOptions {
+  std::optional<Ort::SessionOptions> _sessionOptions;
+}
 
-      outputs[outputNameArray[i]] = outputValue;
+#pragma mark - Public
+
+- (nullable instancetype)initWithError:(NSError**)error {
+  self = [super init];
+  if (self) {
+    try {
+      _sessionOptions = Ort::SessionOptions{};
+    } catch (const Ort::Exception& e) {
+      ORTSaveExceptionToError(e, error);
+      self = nil;
     }
+  }
```

Comment:
Should there be an additional API END macro that handles this pattern of setting self to nil instead of returning failure_return_value?

Reply:
reworked init methods to be more consistent and use the same macro as other methods returning nullable values